src/java/org/apache/cassandra/index/sai/disk/v1/OffsetFactory.java

```diff
@@ -0,0 +1,157 @@
+
+import java.io.IOException;
+
+import org.apache.cassandra.index.sai.SSTableQueryContext;
+
+public class OffsetFactory implements LongArray.Factory
+{
+    private final LongArray.Factory wrapped;
+    private final long idxOffset;
+
+    OffsetFactory(LongArray.Factory wrapped, long idxOffset)
+    {
+        this.wrapped = wrapped;
+        this.idxOffset = idxOffset;
+    }
+
+    @Override
+    public LongArray open()
+    {
+        return new OffsetLongArray(wrapped.open(), idxOffset);
+    }
+
+    @Override
+    @SuppressWarnings({"resource", "RedundantSuppression"})
+    public LongArray openTokenReader(long segmentRowId, SSTableQueryContext context)
+    {
+        // apply segment offset so that `LongArray.findTokenRowId` will start searching tokens from current segment.
+        LongArray reader = wrapped.openTokenReader(segmentRowId + idxOffset, context);
+        return new TokenLongArray(context, reader, idxOffset);
+    }
+
+    private static class OffsetLongArray implements LongArray
+    {
+        private final LongArray wrapped;
+        protected final long idxOffset;
+
+        OffsetLongArray(LongArray wrapped, long idxOffset)
+        {
+            this.wrapped = wrapped;
+            this.idxOffset = idxOffset;
+        }
+
+        /**
+         * Get value at {@code idx}.
+         */
+        @Override
+        public long get(long idx)
+        {
+            return wrapped.get(toSSTableRowId(idx));
+        }
+
+        @Override
+        public long findTokenRowID(long value)
+        {
+            // Subtract the segment offset from the global row ID to provide a segment row ID for the caller:
+            return toSegmentRowId(wrapped.findTokenRowID(value));
+        }
+
+        /**
+         * Get array length.
+         */
+        public long length()
+        {
+            return wrapped.length();
+        }
+
+        @Override
+        public void close() throws IOException
+        {
+            wrapped.close();
+        }
+
+        protected long toSSTableRowId(long segmentRowId)
+        {
+            return segmentRowId + idxOffset;
+        }
+
+        protected long toSegmentRowId(long sstableRowId)
+        {
+            return sstableRowId - idxOffset;
+        }
+    }
+
+    /**
+     * Cache the prev token value and prev sstable row id pair, and share it between different indexed columns in the
+     * same query.
+     */
+    static class TokenLongArray extends OffsetLongArray
+    {
+        private final SSTableQueryContext context;
+
+        TokenLongArray(SSTableQueryContext context, LongArray wrapped, long idxOffset)
+        {
+            super(wrapped, idxOffset);
+            this.context = context;
+        }
+
+        @Override
+        public long get(long segmentRowId)
+        {
+            long sstableRowId = toSSTableRowId(segmentRowId);
+            if (sstableRowId == context.prevSSTableRowId)
+            {
+                return context.prevTokenValue;
```

Comment:
Untested code path?

Reply:
This is all a bit strange to me. We have now moved `OffsetFactory` to `test` because that is the only place it is used. We then have `TokenLongArray` that is only used in `OffsetFactoryTest`. So it appears that we have classes that are only being used to test a class that is a test utility.

I'm going to look closer at this whole class and see if it is needed at all. I'm inclined to delete it and make any tests that use it, use non-test classes in it's place.

Reply:
Ok, I looked at it and it's ridiculous. The test in `NumericValuesTest` that used `OffsetFactory` was, again, only testing functionality in `OffsetFactory`. Since I'm beefing up the testing around the `AbstractBlockPackedReader` classes anyway, I'm going to get rid of `OffsetFactory` and it's associated usage.

Reply:
Beautiful